### DISCUSSION TITLE
JWT Token configuring expiration and/or React Native Guidance

### DISCUSSION THREAD

**Discussion by wollerman** [nhost/nhost#1890] [General]
I had a question related to auth specifically for our native builds. Our goal is to keep users signed in when they come back to our app at a later time, e.g. maybe a few hours. Right now, once a native app is closed, it seems like we the token is no longer available. 

So for React Native, is it possible to configure the token to be stored between sessions? I saw another discussion about how tokens are only configured in memory, so that might prevent this.

Relatedly, if we are able to store tokens, it seems like there is a 15 minute refresh window. Is it possible to configure this for our application? In the native app, it's cumbersome to log in every 15 minutes and we'd prefer to keep the session valid for longer.

Thanks for any feedback and help!

**Comment by wollerman** [comment]
So after thinking about this and digging through some types, the `closing and opening` losing token is fixed by configuring the native store. I also realized I left out details that I'm using expo.

I found the `clientStorageType` which led me to install https://docs.expo.dev/versions/latest/sdk/securestore/. And now I've set my `nhost` client to be configured as follows:

```javascript
const nhost = new NhostClient({
  subdomain: configForEnv.nhost.subdomain,
  region: configForEnv.nhost.region,
  autoSignIn: true,
  autoRefreshToken: true,
  clientStorageType: Platform.OS === "web" ? "web" : "expo-secure-storage",
  clientStorage: Platform.OS === "web" ? undefined : SecureStore,
});
```

This seems to work as expected. But if anyone sees an issue please let me know!

And that just leaves the other question which is: will this not refresh the JWT after 15 minutes if I completely close the app? And is there a way to extend that time?

**Reply by wollerman** [reply]
So to partially answer my own question, I updated my app with the above config and tested out opening ever few hours. After 24 hours I'm still able to fully log in.

So I guess only my question about configuration and actual duration is left if anyone has any insight.

**Reply by fmanzali** [reply]
That is expected, I think. The `clientStorage` you set is used for storing the [refresh token](https://docs.nhost.io/authentication/tokens#refresh-token), which is long lived. This is, effectively, the "session": the refresh tokens associated to each user are stored on the database. 

The access token, instead, is the one that expires every 15 minutes, and is used for authenticating requests. But while the refresh token is active, the client will automatically request a new access token whenever needed. If you close the app for more than 15 minutes, but keep the refresh token in the store, a new access token will simply be requested when you reopen the app.

I would suggest not to modify the access token expiration: it is supposed to be short because of security reasons. An access token can't be revoked - so if someone steals it, you have no way to force them to log out.
Anyways, looking at the code, I see [here](https://github.com/nhost/hasura-auth/blob/b2a1dbb8fc6d72cf070e19a4b8e4c9b01fa12655/src/utils/jwt/generate.ts#LL24C55-L24C55) that the expiration time is set by an environment variable - so you can try to do so [from the dashboard](https://docs.nhost.io/platform/environment-variables) (no guarantees, I have not tried it).

Instead, you may want to modify the expiration of the refresh token. Similarly, this is configured in a different [env variable](https://github.com/nhost/hasura-auth/blob/b2a1dbb8fc6d72cf070e19a4b8e4c9b01fa12655/src/utils/refresh-token.ts#LL37C57-L37C57), so you can try that.

**Reply by wollerman** [reply]
Ah, seems like I was misunderstanding. Thanks for the explanation! This makes a lot of sense!

**Reply by wollerman** [reply]
@fmanzali following up, in the nhost console it seems I can't override the refresh token value:
<img width="376" alt="image" src="https://github.com/nhost/nhost/assets/1610241/0157357d-bccf-47aa-9db8-21919d1fb14c">

Is there somewhere else this should be set? I'm happy to also update the docs depending on how this should be managed.

Thanks!

**Reply by fmanzali** [reply]
Unfortunately I think they may not be configurable - see #498. I see they are grouped as "System Environment Variables" which are not directly editable. Perhaps someone from the Nhost team could help you with that.

Anyways, what is your requirement for sessions? Perhaps there is a workaround to handle that without editing the global configuration.

**Reply by wollerman** [reply]
Ah interesting.

The reason I'm trying out different workarounds is I've noticed an issue where my refresh token isn't actually valid for 30 days. If I log in now and inspect the `nhostRfreshTokenExpiresAt`, it expires in 12 hours. So I'm trying to figure out what's different about my setup to cause the short lived refresh.

And I can confirm this by looking at the nhost console `auth.refresh_tokens` row which has a value of `expires_at` in 12 hours.

**Reply by fmanzali** [reply]
I think you may be in luck, the new [config update](https://nhost.io/blog/config) seems to expose all the env variables you need for changing the access/refresh token lifetime. 
For instance, for the access token, you may try editing this in `nhost.toml` (after upgrading the CLI to v1):

```toml
[auth.session]
[auth.session.accessToken]
expiresIn = {VALUE_IN_SECONDS}
```

And for the refresh token:
```
[auth.session.refreshToken]
expiresIn = {VALUE_IN_SECONDS}
```

I have also inspected my setup, and I see that also for me the `nhostRefreshTokenExpiresAt` is set to 12 hours in the future. So I looked at it a bit more, and saw that the default refresh token lifetime from `nhost.toml` is `43200` seconds, or `12 hours`. So this makes sense, the [docs](https://docs.nhost.io/authentication/tokens#refresh-token) are simply not updated.

In fact, I guess that 30 days was the previous default value. If you use cookies for storing the tokens, [this code](https://github.com/nhost/nhost/blob/ef05d698897016312fa1495bbc2402086a321a71/packages/hasura-auth-js/src/local-storage.ts#L85-L87) is executed, and here you see the 30 days expiration. But the actual value of `nhostRefreshTokenExpiresAt` is computed from the env variable, as it should be.

**Reply by dbarrosop** [reply]
yes, that is correct. Those options should allow you to tweak those timers.

**Reply by wollerman** [reply]
Amazing; yep this is exactly what I'm looking for. I'll give the upgrade a try!

**Reply by wollerman** [reply]
So I've gone through the upgrade and updated the the `nhost.toml` to have:
```
[auth.session.refreshToken]
expiresIn = 2592000 # 30 days
```
After deploying I'm still seeing tokens in web storage for 12 hours. e.g. logging in right now and inspecting I see 
![image](https://github.com/nhost/nhost/assets/1610241/fc3939e2-1da3-4697-9f90-106da3291cec)

I can see it picked up the new config in the nhost dashboard. I see output for the deploy:

```
07:24:23:
07:24:23: Starting deployment
07:24:23: Cloning repo <redacted> (#b03a4b3)
07:24:25: Detected new config file
07:25:04: App's configuration has been updated
07:25:04: Applying database migrations
07:25:07: Database migrations applied successfuly
07:25:07: Applying metadata
07:25:11: Metadata applied successfully
07:25:11: Deploying functions
07:25:27: Functions deployment completed
07:25:27: Reloading metadata
07:25:30: Deploying custom email templates
07:25:31: Deployment completed with status DEPLOYED
```

I've even set it to a very large number like `31536000` which should be 365 days, unless my math is off.

Is there something else to modify to change this value? Or is it not fully supported to be changed via the toml config? Or perhaps the value has been updated and what I'm seeing in storage doesn't affect the refresh token as I'm expecting?

I'll have to wait at least until this evening to test that it's made it past the 12 hours and report back :)

**Reply by dbarrosop** [reply]
What is your subdomain so I can double check?

**Reply by wollerman** [reply]
wtevrmklpcuquyeuvzpy

**Reply by dbarrosop** [reply]
I reviewed everything and it looks fine. Maybe you were reusing an old session with a token that still had the old refresh token expiration time? You could try logging out and then logging in again and see if that fixes it. To be sure, you could also delete all the the existing tokens in refresh_token table.
